dandi/download.py

```diff
@@ -309,13 +312,33 @@ def download_generator(
                     lock=lock,
                 )
 
+            # If exception is raised we might just raise it, or yield
+            # an error record
+            gen = {
+                "raise": _download_generator,
+                "yield": _download_generator_guard(path, _download_generator),
+            }[on_error]
+
             if yield_generator_for_fields:
-                yield {"path": path, yield_generator_for_fields: _download_generator}
+                yield {"path": path, yield_generator_for_fields: gen}
             else:
-                for resp in _download_generator:
+                for resp in gen:
                     yield dict(resp, path=path)
 
 
+def _download_generator_guard(path: str, generator: Iterator[dict]) -> Iterator[dict]:
+    try:
+        for resp in generator:
+            yield dict(resp, path=path)
+    except Exception as exc:
+        lgr.error("Caught while downloading %s", path, exc_info=exc)
+        yield {
+            "path": path,
```

Comment:
This line should be eliminated, as `"path"` is already set on lines 323 and 326.

Reply:
```suggestion
```